Control flow based type guards

ivogabe opened the pull request:
As discussed in #5168, here's my PR that adds flow based type guards. This is still work in process. The compiler can compile itself, but one of the tests doesn't terminate.

Related issues: #1764, #1769, #2388

Here's a quick overview of the approach I took. During binding, every node gets references to flow markers that preceded the node in the control flow. A flow marker is either a type guard or some node, such as an identifier or a while loop. This represents the control flow graph.

In `checker.ts`, the type of a variable is calculated as the union of the types after the previous occurrences. The type after a node is in case of an assignment narrowed by the assigned type (see below) and otherwise the same as before the node.

At an assignment `x = y`, where `y` can be any expression, the following happens. If the type of `x` is a union type, the constituent parts are filtered. The local type of `x` will be the union of all constituent parts to which `y` is assignable. If this yields an empty union type, the initial type of `x` is used instead. If `x` is not a union type, the local type after the assignment will be the initial type.

Since the incoming flow of a `catch` block can come from anywhere in the `try` block, I decided to fall back to the initial type as a precaution. I can relax that rule if desired, for instance by looking at assignments in the `try` block.

Todo:
- [x] Fix test where compiler doesn't terminate
- [x] Add tests
- [x] Clean up & comments
- [x] Fix bug: `let x: string | number = ""` doesn't narrow to `string`, but `x = ""` does.
- [x] Fix bug: with `x = ""; for (;x;) { x; }`, narrowing is removed inside the for (because of recursion), but `x = ""; for(;;) { x; }` works as expected. 

~~Questions:
1. Not every `Node` has an `id` property (at runtime it is `undefined`). Is that correct or a bug? I currently use `flowIndex` (which I will remove later on) in `getPreviousOccurencies`, but `id` would be better.
2. One test currently doesn't terminate. The process aborts because it's out of memory (even when I increase the old space to 8 GB). Given that the call stack contains `createFileDiagnostic`, I think the compiler is trying to create an infinite amount of diagnostics.. Does someone have an idea why that's happening?~~

weswigham commented:
> 1. Not every Node has an id property (at runtime it is undefined). Is that correct or a bug? I currently use flowIndex (which I will remove later on) in getPreviousOccurencies, but id would be better.

Use the function `getNodeId` to get node id's - if the node has yet to be assigned one, this method assigns it one.

weswigham commented on src/compiler/utilities.ts:
```diff
@@ -1405,7 +1405,7 @@ namespace ts {
 
     // True if the given identifier, string literal, or number literal is the name of a declaration node
-    export function isDeclarationName(name: Node): name is Identifier | StringLiteral | LiteralExpression {
+    export function isDeclarationName(name: Node): name is (Identifier & { __weakTypeGuard: void; }) | StringLiteral | LiteralExpression {
```
I don't see this brand applied anywhere - is there a reason for this?

ivogabe replied:
If a type guard returns false, it is expected that the argument is not of the specified type. However, `isDeclarationName` can return false when `name` is an identifier. This wasn't an issue, as `isDeclarationName` was never used in an if statement with an else. However, with control flow based type guards, this wasn't working anymore. The then-block of an if statement contained a return, thus the statements after the if were the else branch of the if. It might not be the best way to solve this, but it's working..

weswigham commented:
After a bit of sniffing with your code, I think the (or one of the?) test(s) which hangs is `tests/cases/conformance/parser/ecmascript5/parserRealSource7.ts` (which I think is a chunk of code form the original compiler). The `preCollectVarDeclTypes` function within the file is enough (on its own) to make the test timeout (though not also hit the memory ceiling) - adding a definition for `hasFlags` to the test then allows it to complete inside the given time. I think symbols with unknown types are making the analysis take an unreasonably long time/memory/recursive depth.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -6852,81 +6852,135 @@ namespace ts {
             }
         }
 
+        interface PreviousOccurency {
```
This type isn't used anywhere, AFAIK.

ivogabe replied:
You're right, I forgot to remove it during refactoring.

ivogabe commented:
@weswigham Thanks for your help! I have removed `flowIndex` and I'm now using `getNodeId`. I have removed the unused interface. I will take a look at the test later today or this week.

ivogabe commented on src/compiler/types.ts:
```diff
@@ -472,10 +483,19 @@ namespace ts {
         text: string;                                  // Text of identifier (with escapes converted to characters)
         originalKeywordKind?: SyntaxKind;              // Original syntaxKind which get set so that we can report an error later
+        /* @internal */ narrowingState?: NarrowingState;
+        /* @internal */ localType?: Type;
```
Should this be here or in `NodeLinks`?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13376,12 +13474,12 @@ namespace ts {
         }
 
-        function checkIteratedTypeOrElementType(inputType: Type, errorNode: Node, allowStringInput: boolean): Type {
+        function checkIteratedTypeOrElementType(inputType: Type, errorNode: Node, allowStringInput: boolean, showError: boolean): Type {
```
Nit: we usually call these "reportError`

ahejlsberg commented:
@ivogabe This definitely looks interesting.

I haven't had a chance to review the code yet, but I'm curious how it affects performance. For example, when you compile the compiler with `-diagnostics` how do the numbers differ from the regular compiler?

Also, do you reason about effects of closure over outer local variables from nested functions? For example, calls to the same local function can have different effects depending on parameters. Obviously not an easy problem to solve.

ivogabe commented:
@ahejlsberg Thanks! I haven't tried `-diagnostics`, but I'm sure it's too slow at the moment. I have some ideas to make it faster that I haven't implemented.

What do you mean exactly with nested functions? Both the old (current) and my implementation work this way:
```typescript
let x: string | number;
function setX() { x = 42; }
if (typeof x === "string") {
  setX();
  // x: string
}
```

With my implementation, narrowing happens only within a function, so on the marked position `x` would not be narrowed. This is because I cannot create a graph with these function calls.
```typescript
if (typeof x === "string") {
  let y = () => {
    // x: string | number;
  };
}
```

ahejlsberg commented:
@ivogabe I was wondering about code like this:

```Typescript
let x: string | number;
initialize(true);
// Is x considered assigned here and what is its type?
function initialize(b: boolean) {
    if (b) {
        x = 42;
    }
}
```

ivogabe commented:
@ahejlsberg Ah, in that case `x` would not be narrowed after the function call, on the marked spot. In the function definition it would be narrowed after the assignment of course.

ivogabe commented:
I found the issue with the test. I thought that the test didn't terminate, but it just took way too long and it used too much memory. In the following code, `2^n` paths exist from the first use of `x` to the second `x` on the last line:
```typescript
let x: string | number;
x = 42;
if (...) {}
if (...) {}
...
if (...) {}
x
```
The old implementation would try all paths (`2^n`, with `n` the amount of if statements). The type of `x` is now cached before each if statement, which means that `2n` paths are now analyzed. That was implemented in 2c57ee5290cfa5cc298f22a1bae2c4d8c000c430. I have also added some more caching, which decreased compilation time too.

Narrowing of identifiers is working now, only destructuring doesn't work correctly.

@ahejlsberg I get these numbers when I enable `--diagnostics`:
This branch:
```
Files:              40
Lines:           82685
Nodes:          396459
Identifiers:    148455
Symbols:       1563523
Types:           15469
Memory used:   343507K
I/O read:        0.01s
I/O write:       0.06s
Parse time:      0.67s
Bind time:       0.66s
Check time:      3.65s
Emit time:       1.13s
Total time:      6.10s
```
Master:
```
Files:              40
Lines:           82331
Nodes:          394293
Identifiers:    147659
Symbols:       1558322
Types:           15418
Memory used:   322984K
I/O read:        0.01s
I/O write:       0.06s
Parse time:      0.67s
Bind time:       0.52s
Check time:      2.22s
Emit time:       1.05s
Total time:      4.46s
```
That's an increase of 37%. Should I do some additional optimizations?

ivogabe commented:
With commit 09e4302bbcd4902b1e238859a343703a0a6cd4b7, I have reduced it to 20%:
```
Files:              40
Lines:           82838
Nodes:          397641
Identifiers:    148909
Symbols:       1565494
Types:           15637
Memory used:   346028K
I/O read:        0.01s
I/O write:       0.04s
Parse time:      0.63s
Bind time:       0.64s
Check time:      3.05s
Emit time:       1.05s
Total time:      5.36s
```

ivogabe commented:
It took a while, but I think most things are working now. I have updated the PR for the dotted names type guards and this type guards. `null` and `undefined` worked without extra work. I haven't used the control flow graph for variable assignment checking, but that can be easily added later on.

I think that two minor problems: handling of a try/catch block and `getTypeOfSymbolAtLocation`. After a try/catch, narrowing breaks and types of variables are set to their initial types. Better would be to use the type before the try if the variable is not assigned within the try, otherwise the initial type; or to take the union of the type before the try with all types of assignments in the try/catch.

The following code compiles, but gives the wrong type in VSCode, with this custom build. When I hover the last `x`, it says `string | number`. I suspect this has something to do with `getTypeOfSymbolAtLocation`, but I'm not sure what's happening. This seems to fail with double type guards. The tests show this too, in `tests/baselines/reference/typeGuardsInIfStatement.types`.
```typescript
let x: string | boolean | number;
if (typeof x === "string") {}
else if (typeof x === "boolean") {}
else x.toExponential;
```

Since most work has been finished, can someone review my code?

jods4 commented:
For `try...catch` and `try..finally` it's very useful to keep the narrowed type of symbols that do not change type in the block, including those that are assigned, so that the following works:
```typescript
let x: number | null;
x = 4; // from here x: number
try {
  x = 7;
  throw new Error();
} 
finally {
  x.toString(); // <-- should be ok, it would be annoying if x reverts to number|null
}
```

For symbols, which change type inside the block I am unsure whether it's worth the effort. The union of all their types across the block would be correct but it can get hard to reason about, for IMHO little benefit. I think reverting to their initial type would be OK for a first release, it can always be improved later if people have motivating examples of code where it would be nice.

mhegazy commented:
this functionality should be handled by https://github.com/Microsoft/TypeScript/pull/8010.

Big thanks to @ivogabe for sharing this work. it has been the inspiration and guide for the work in https://github.com/Microsoft/TypeScript/pull/8010. thanks again @ivogabe!
